spinalcordtoolbox/labels.py

```diff
@@ -0,0 +1,473 @@
+    :param img: source image
+    :param coordinates: list of Coordinate objects (see spinalcordtoolbox.types)
+    :returns: labeled source image
+    """
+    out = _add_labels(img.copy(), coordinates)
+
+    return out
+
+
+def _add_labels(img: Image, coordinates: Sequence[Coordinate]) -> Image:
+    """
+    Given an image and list of coordinates, add the labels to the image and return it.
+    :param img: source image
+    :param coordinates: list of Coordinate objects (see spinalcordtoolbox.types)
+    :returns: labeled source image
+    """
+
+    for i, (x, y, z, v) in enumerate(coordinates):
+        if len(img.data.shape) == 3:
+            img.data[int(x), int(y), int(z)] = v
+        elif len(img.data.shape) == 2:
+            if z != 0:
+                raise ValueError(f"2D coordinates should have a Z value of 0! Current value: {z}")
+            img.data[x, y] = v
+        else:
+            raise ValueError(f"Data should be 2D or 3D. Current shape is: {img.data.shape}")
+
+        logger.debug(f"Label #{i}: {x}, {y}, {z} --> {v}")
+
+    return img
+
+
+def create_labels_along_segmentation(img: Image, labels: Sequence[Tuple[int, int]]) -> Image:
+    """
+    Create an image with labels defined along the spinal cord segmentation (or centerline).
+    Input image does **not** need to be RPI (re-orientation is done within this function).
+    :param img: source segmentation
+    :param labels: list of label tuples as (z_value, label_value)
+    :returns: labeled segmentation (Image)
+    """
+    og_orientation = img.orientation
+
+    if og_orientation != "RPI":
+        img.change_orientation("RPI")
+
+    out = zeros_like(img)
+
+    for idx_label, label in enumerate(labels):
+        z, value = label
+
+        # update z based on native image orientation (z should represent superior-inferior axis)
+        coord = Coordinate([z, z, z])  # since we don't know which dimension corresponds to the superior-inferior
+
+        # axis, we put z in all dimensions (we don't care about x and y here)
+        _, _, z_rpi = coord.permute(img, 'RPI')
+
+        # if z=-1, replace with nz/2
+        if z == -1:
+            z_rpi = int(np.round(out.dim[2] / 2.0))
+
+        # get center of mass of segmentation at given z
+        x, y = ndimage.measurements.center_of_mass(np.array(img.data[:, :, z_rpi]))
+
+        # round values to make indices
+        x, y = int(np.round(x)), int(np.round(y))
+
+        # display info
+        logger.debug(f"Label # {idx_label}: {x}, {y}. {z_rpi} --> {value}")
+
+        if len(out.data.shape) == 3:
+            out.data[x, y, z_rpi] = value
+        elif len(out.data.shape) == 2:
+            if z != 0:
+                raise ValueError(f"2D coordinates should have a Z value of 0! Current value: {coord.z}")
+
+            out.data[x, y] = value
+
+    if out.orientation != og_orientation:
+        out.change_orientation(og_orientation)
+        img.change_orientation(og_orientation)
```

Comment:
`img` is used in the computation of `out`. If the orientation is not `RPI` it is re-oriented, therefore I am restoring the original orientation. If I didn't that would cause a side-effect for the caller.

The alternative is to make a copy of `img` and use that in the processing. In this specific case I opted to avoid the image copy for performance reasons. I would not want to sacrifice readability for (a probably small) RAM performance improvement however. 

Do you think this negatively affects the readability of the functions?

Reply:
> Do you think this negatively affects the readability of the functions?

To me it does, but only a little. It's a small trade-off either way, so I trust your judgement here. :)